### ISSUE TITLE
Feature request: difference and intersection operations for Path2 vs Region2

### ISSUE THREAD

**Issue by edwinRNDR** [lacuna/artifex#2]
Would it be possible to add difference and intersection operations for open shapes (Path2) versus closes ones (Region2)?

**Comment by ztellman** [comment]
It’s not immediately clear to me what the semantics of those operations would be for a non-enclosed shape (typically they denote operations on a set). Can you provide some examples?

**Comment by edwinRNDR** [comment]
![Screenshot 2020-09-14 at 09 30 32](https://user-images.githubusercontent.com/983309/93056738-24f31500-f66d-11ea-9e4c-40004159a627.png)

Here you see the intersection and difference operators for Path2 (pink) vs Region2 (cyan-ish) and the result in black. They are still operators that work on a set, but in this case the set is not one of polygon area.

**Comment by ztellman** [comment]
Ah, I see, I was thinking `(Path, Path) => Path`, not `(Path, Region) => Path`.  Yes, that should be pretty straightforward, it's just the initial steps of the `(Region, Region) => Region` logic.  Some of the consistency checks we use for the region operations (see https://ideolalia.com/2018/08/28/artifex.html) won't be possible here, so there may be more artifacts for pathological inputs, but that's difficult to quantify. 

I'll update once I have some code thrown together for this.

**Comment by edwinRNDR** [comment]
Yes, this is what I had in mind! Thank you, looking forward to seeing that code.

**Comment by hamoid** [comment]
Would be very nice to have also have splitting on intersections (closed-closed, closed-open and open-open contours). 

Here an animation showing one of the cases: https://openrndr.discourse.group/t/doodles-wall-c/51/73?u=abe

The result is always a list of open contours, no matter if the sources are closed or open. In my code I have two arguments, a target shape and a cutter shape.

When a open contour crosses or touches a closed contour in a single point the only change that takes place on the target closed contour is that the first vertex becomes the intersection point (so the points are sorted). This is useful for animation (if I want to grow the contour starting at the cutting point) or effects like in this doodle, where I cut a series of parallel lines or concentric circles and then apply some forces to the resulting lines.
![cut](https://user-images.githubusercontent.com/108264/93750835-6ac75480-fbfc-11ea-8c4d-a3fbf998b989.png)

**Comment by ztellman** [comment]
I think the general form of this would be a `Path2.split(Curve2[] splitters) => Iterable<Path2>` method, which would return the original path split wherever it intersects the provided curves.  For the intersection/difference method, you can just test a point within each of the resulting paths against the region.

Does that seem reasonable?

**Comment by edwinRNDR** [comment]
That sounds fair. I see there is a `Region2.test()`, that would make figuring out if a path is inside or outside simple enough.

**Comment by ztellman** [comment]
I'd probably still have explicit `intersection`/`difference` methods, because empirically I've found it's best to do the `test` somewhere other than the midpoint of a curve, because that often will be near the edge of the other intersecting shape (I use `1/e` instead), and that's not something I'd expect a user of the library to have to know or worry about.
